src/theory/quantifiers/sygus/sygus_reconstruct.h

```diff
@@ -0,0 +1,302 @@
+/*********************                                                        */
+/*! \file sygus_reconstruct.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Abdalrhman Mohamed
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief utility for reconstructing terms to match a grammar
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef CVC4__THEORY__QUANTIFIERS__SYGUS_RECONSTRUCT_H
+#define CVC4__THEORY__QUANTIFIERS__SYGUS_RECONSTRUCT_H
+
+#include <map>
+#include <vector>
+
+#include "theory/quantifiers/sygus/rcons_obligation_info.h"
+#include "theory/quantifiers/sygus/rcons_type_info.h"
+
+namespace CVC4 {
+namespace theory {
+namespace quantifiers {
+
+/** SygusReconstruct
+ *
+ * This class implements an algorithm for reconstructing a given term such that
+ * the reconstructed term is equivalent to the original term and its syntax
+ * matches a specified grammar.
+ *
+ * Goal: find a term g in sygus type T that is equivalent to builtin term t.
+ *
+ * rcons(t, T) returns g
+ * {
+ *   // a list of triples to reconstruct into Sygus type T, where
+ *   // a labeled term is of the form (k, t, s), where k is a skolem of type T,
+ *   // t is a builtin term of the type encoded by T, and s is a possibly null
+ *   // sygus term of type T representing the solution.
+ *   Obs<T>
+ *
+ *   // We define Term<k> and Sol<k> for each skolem k such that Obs<T> contains
+ *   // entries (k, Term<k>, Sol<k>)
+ *
+ *   CandSols<k> : a list of possible solutions, for each label k. Whenever
+ *                 there is a successful match, it's added to this list.
+ *   Pool<T> : a list of enumerated terms for each sygus type
+ *
+ *   let k_0 be a fresh skolem of sygus type T
+ *
+ *   Obs<T> += {(k_0, t, null)}
+ *
+ *   Pool<T> = {}
+ *
+ *   while Sol<k_0> == null
+ *     // enumeration phase
+ *     for each T
```

Comment:
What T do we consider?

Reply:
Updated to ` for each subfield type T of T_0`. Not sure if that's more intuitive though.